BUGFIX: Memory Leak in NetscriptPorts

jeek opened the pull request:
Prior to making this change, a script that uses a non-trivial amount of Netscript Ports once each and then never uses them again would cause the empty ports to hang around, taking up a growing amount of memory. This patch deallocates empty ports from the NetscriptPorts map.

Tested by running my script that uses a non-trivial number of ports once each and watching the Memory tab in the debug console. Prior to the patch, it would grow until game crashed, at a rate of about 2MB/s. After the patch, memory usage grew to about 300MB and stayed there.

jeek committed:
BUGFIX: Memory Leak in NetscriptPorts

Snarling committed:
Change Port to a class

* Should reduce performance issues from player initializing too many ports.
* Also included several other minor changes (where "port" was referring to a number and not a port, changed to portNumber, port numbers are now initially verified as positive integers / no rounding, etc)

Snarling commented:
@jeek When you get a chance, can you test the new changes I just pushed affect your memory usage with high ports usage? If this doesn't work satisfactorily I have one more idea that may work without breaking portHandle references.

Snarling committed:
portHandle changes and deletePortReference

* After some benchmarking, private class fields are really slow. Changed data and resolvers to public class fields (time to initialize 2e6 ports with ns.readPort was almost 30x faster without private class fields).
* Added deletePortReference netscript function, it's undocumented but it will delete a member of NetscriptPorts and allow garbage collection.
* Changed the way portHandles work, to still hide the underlying data and resolvers fields and persist pointing to the same port number even if the port gets deleted. nextWrite still breaks on port deletion which can't be avoided, so that's one reason port deletion is not automatic and requires manual management.

Snarling committed:
Auto clean up ports

Back to a solution that auto cleans up the ports, but with no impact to portHandles. Reorganized basically all of the ports stuff.

Snarling committed:
Port doc fixes

* Doc no longer says 20 ports max.
* Updated a comment
* Correct type for NetscriptPorts keys
